Central transitive dependencies should be considered only for root nodes

marcin-krystianc opened the pull request:
<!-- DO NOT MODIFY OR DELETE THIS TEMPLATE. IT IS USED IN AUTOMATION. -->
## Bug

<!-- Search https://github.com/NuGet/Home/issues, and create one if you can't find a suitable issue. -->
<!-- Paste the full link, like https://github.com/nuget/home/issues/1000. GitHub will render is neatly. -->
Fixes: https://github.com/NuGet/Home/issues/11760

Regression? Last working version: -

## Description
<!-- Add details about the fix. Include any information that would help the maintainer review this change effective. -->

After this change, the DependencyWalker will check whether transitive dependency eclipses or downgrades the current library only when that transitive dependency is at the top-level node.

With this change, the example from the linked [issue](https://github.com/NuGet/Home/issues/11760) fails with downgrade errors (GOOD):
```
dotnet.exe restore --force
  Determining projects to restore...
  The project D:\workspace\TestSolutions\MissingDowngradeCentralisedLib\Lib1\Lib1.csproj is using CentralPackageVersionManagement, a NuGet preview feature.
  The project D:\workspace\TestSolutions\MissingDowngradeCentralisedLib\App\App.csproj is using CentralPackageVersionManagement, a NuGet preview feature.
D:\workspace\TestSolutions\MissingDowngradeCentralisedLib\App\App.csproj : error NU1605: Detected package downgrade: Newtonsoft.Json from 12.0.2 to 12.0.1. Reference the package directly from the project to select a different version.  [D:\workspace\TestSolutions\MissingDowngradeCentralisedLib\Solution.sln]
D:\workspace\TestSolutions\MissingDowngradeCentralisedLib\App\App.csproj : error NU1605:  App -> Lib1 -> Lib2 -> Newtonsoft.Json (>= 12.0.2)  [D:\workspace\TestSolutions\MissingDowngradeCentralisedLib\Solution.sln]
D:\workspace\TestSolutions\MissingDowngradeCentralisedLib\App\App.csproj : error NU1605:  App -> Newtonsoft.Json (>= 12.0.1) [D:\workspace\TestSolutions\MissingDowngradeCentralisedLib\Solution.sln]
```


## PR Checklist

- [x] PR has a meaningful title
- [x] PR has a linked issue.
- [x] Described changes

- **Tests**
  - [x Automated tests added
  - **OR**
  <!-- Describe why you haven't added automation. -->
  - [ ] Test exception
  - **OR**
  - [ ] N/A <!-- Infrastructure, documentation etc. -->

- **Documentation**
  <!-- Please link the PR/issue if appropriate -->
  - [ ] Documentation PR or issue filled
  - **OR**
  - [x] N/A

marcin-krystianc committed:
Central transitive dependencies should be considered only for root nodes

marcin-krystianc committed:
Central transitive dependencies should be considered only for root nodes

marcin-krystianc commented on src/NuGet.Core/NuGet.ProjectModel/LockFile/LockFileFormat.cs:
```diff
@@ -843,7 +843,7 @@ private static List<CentralTransitiveDependencyGroup> ReadProjectFileTransitiveD
                     NuGetFramework framework = NuGetFramework.Parse(frameworkPropertyName);
                     var dependencies = new List<LibraryDependency>();
 
-                    JsonPackageSpecReader.ReadCentralTransitveDependencyGroup(
```
Just fixing a typographic mistake in an internal method.

jeffkl commented on src/NuGet.Core/NuGet.DependencyResolver.Core/Remote/RemoteDependencyWalker.cs:
```diff
@@ -306,6 +306,10 @@ public async Task<GraphNode<RemoteResolveResult>> WalkAsync(LibraryRange library
             {
+                // Central transitive dependencies should be considered only for root nodes
+                if (!isRoot && d.ReferenceType == LibraryDependencyReferenceType.None)
+                    continue;
```
```suggestion
                {
                    continue;
                }
```

jeffkl reviewed (commented):
Thanks, looks good with one small comment.  I had to rebase your branch and force push so be sure to pull it next time you work on it.

marcin-krystianc committed:
Update src/NuGet.Core/NuGet.DependencyResolver.Core/Remote/RemoteDependencyWalker.cs

Co-authored-by: Jeff Kluge <jeffkl@microsoft.com>

marcin-krystianc commented:
@jeffkl https://github.com/NuGet/NuGet.Client/pull/4611/commits/161253062d48bb71bb42929c51688025fd6f89e0 has some test failures. Are these legitimate issues or just a flakiness?

jeffkl commented:
> @jeffkl https://github.com/NuGet/NuGet.Client/commit/161253062d48bb71bb42929c51688025fd6f89e0 has some test failures. Are these legitimate issues or just a flakiness?

Yeah sorry we're having some troubles with CI today, I'll work on it.

nkolev92 commented on src/NuGet.Core/NuGet.DependencyResolver.Core/Remote/RemoteDependencyWalker.cs:
```diff
@@ -306,6 +306,12 @@ public async Task<GraphNode<RemoteResolveResult>> WalkAsync(LibraryRange library
             foreach (LibraryDependency d in item.Data.Dependencies)
             {
+                // Central transitive dependencies should be considered only for root nodes
+                if (!isRoot && d.ReferenceType == LibraryDependencyReferenceType.None)
```
I'm not sure the isRoot check is correct here. 

This might disable warnings for scenarios like  

```
Project -> A 1.0.0 -> B 1.0.0
                   -> C 1.0.0 -> B 2.0.0
```                          
This should select B 1.0.0 and give you a downgrade warning.

marcin-krystianc replied:
@nkolev92 Is in your example B 1.0.0, a central transitive dependency, i.e. is it defined in `Directory.Packages.props` but there is no explicit PackageReference in A?

nkolev92 replied:
No, this is a scenario without CPM. LibraryDependencyReferenceType.None is what's there for non CPM.

marcin-krystianc replied:
I think it is exactly opposite, `LibraryDependencyReferenceType.None`  is used exclusively for CPM and transitive dependency pinning: https://github.com/NuGet/NuGet.Client/blob/f52a83999070237288d7e1c8ec5934a65eda77ac/src/NuGet.Core/NuGet.ProjectModel/PackageSpecReferenceDependencyProvider.cs#L342-L354 and https://github.com/NuGet/NuGet.Client/blob/f52a83999070237288d7e1c8ec5934a65eda77ac/src/NuGet.Core/NuGet.DependencyResolver.Core/Remote/RemoteDependencyWalker.cs#L486-L496 and https://github.com/NuGet/NuGet.Client/blob/f52a83999070237288d7e1c8ec5934a65eda77ac/src/NuGet.Core/NuGet.DependencyResolver.Core/Remote/RemoteDependencyWalker.cs#L486-L496

So I sustain that this condition is right and doesn't affect non-transitive pinning scenarios.

nkolev92 commented on test/NuGet.Core.Tests/NuGet.DependencyResolver.Core.Tests/RemoteDependencyWalkerTests.cs:
```diff
@@ -1840,6 +1840,44 @@ public async Task WalkAsync_WithDowngradeInPrunedSubgraph_DoesNotReportDowngrade
+            var context = new TestRemoteWalkContext();
+            var provider = new DependencyProvider();
+            provider.Package("A", "1.0")
+                .DependsOn("B", "1.0")
```
I think `B` needs to be centrally managed here, as B is a top level to A, which effectively means all it's direct deps need to be centrally managed.

marcin-krystianc replied:
`B` is a project here so that is why it is not "centrally managed" as any package would be.

nkolev92 commented on test/NuGet.Core.Tests/NuGet.DependencyResolver.Core.Tests/RemoteDependencyWalkerTests.cs:
```diff
@@ -1840,6 +1840,44 @@ public async Task WalkAsync_WithDowngradeInPrunedSubgraph_DoesNotReportDowngrade
+
+            provider.Package("B", "1.0")
+                .DependsOn("C", "1.0")
+                .DependsOn("D", "3.0", LibraryDependencyTarget.Package, versionCentrallyManaged: true, libraryDependencyReferenceType: LibraryDependencyReferenceType.None);
```
Should this be centrally managed?
In the context of A, this isn't centrally managed, it's just a dependency of B.

marcin-krystianc replied:
`B` is a project that uses CPM, so the package  `D`  dependency should be centrally managed.

nkolev92 replied:
but not in the context of A. 

When A is getting restored, A doesn't care about central management configured by B, it just cares about the declared dependencies (and their versions) from B.

nkolev92 replied:
~but not in the context of A.~

~A and B are restored separately.
When A is getting restored, A doesn't care about central management configured by B, it just cares about the declared dependencies (and their versions) from B.~

This is tricky, as it seems like we don't really recognize the concept of transitively pinned for project 1 but not pinned for project 2.

nkolev92 commented on test/NuGet.Core.Tests/NuGet.DependencyResolver.Core.Tests/RemoteDependencyWalkerTests.cs:
```diff
@@ -1840,6 +1840,44 @@ public async Task WalkAsync_WithDowngradeInPrunedSubgraph_DoesNotReportDowngrade
+            var provider = new DependencyProvider();
+            provider.Package("A", "1.0")
+                .DependsOn("B", "1.0")
+                .DependsOn("D", "1.0", LibraryDependencyTarget.Package, versionCentrallyManaged: true, libraryDependencyReferenceType: LibraryDependencyReferenceType.None);
```
Do we another another level of dependencies here?

If A is the root node, and it's centrally managed, then all it's dependencies need to be centrally managed and are direct.

nkolev92 reviewed (commented):
Sorry for the delay. 

Your explanation seems good!  It's been a while since we've looked at this.

A few qs about the test.

I wouldn't be the worst idea to define an end to end test as well, as the dependency walker tests rely on a *lot* of configuration. 

Example test is ExecuteAsync_WithDowngradesInPrunedSubgraph_DoesNotRaiseNU1605, it's not CPM, but other in the same class are RestoreCommand_CentralVersion_ErrorWhenDependenciesHaveVersion.

marcin-krystianc committed:
Merge remote-tracking branch 'origin/dev' into dev-marcink-20220504-centraltransitive

marcin-krystianc committed:
E2E ExecuteAsync_TransitiveDependenciesFromNonRootLibraries_AreIgnored

marcin-krystianc committed:
Merge remote-tracking branch 'marcink/dev-marcink-20220504-centraltransitive' into dev-marcink-20220504-centraltransitive

# Conflicts:
#	src/NuGet.Core/NuGet.DependencyResolver.Core/Remote/RemoteDependencyWalker.cs
#	test/NuGet.Core.Tests/NuGet.DependencyResolver.Core.Tests/RemoteDependencyWalkerTests.cs

marcin-krystianc commented:
> A few qs about the test.

I think that all these questions are due to confusion about which dependency is a project and which is a package. I've updated the test so now it should be clearer. Please have another look.

> It wouldn't be the worst idea to define an end to end test as well, as the dependency walker tests rely on a lot of configuration.

I implemented `ExecuteAsync_TransitiveDependenciesFromNonRootLibraries_AreIgnored` - it has smililrar configuration complexity to the walker tests though 😞

nkolev92 commented on test/NuGet.Core.Tests/NuGet.Commands.Test/RestoreCommandTests.cs:
```diff
@@ -2544,6 +2544,98 @@ public async Task ExecuteAsync_WithDowngradesInPrunedSubgraph_DoesNotRaiseNU1605
+        ///                -> C 1.0 -> D 2.0
+        /// </summary>
+        [Fact]
+        public async Task ExecuteAsync_TransitiveDependenciesFromNonRootLibraries_AreIgnored()
```
So I created this test with nuget.exe instead (basically, I want to avoid constructing the PackageSpec manually) and here's what I have. 

```cs
/// <summary>
        /// A 1.0 -> D 1.0 (Central transitive)
        ///       -> B 1.0 -> D 3.0 (Central transitive - should be ignored)
        ///                -> C 1.0 -> D 2.0
        /// </summary>
        [Fact]
        public async Task NuGetExeRestore_Test()
        {
            // Arrange
            using var pathContext = new SimpleTestPathContext();

            // Set up solution, project, and packages
            var solution = new SimpleTestSolutionContext(pathContext.SolutionRoot);

            var projectA = SimpleTestProjectContext.CreateNETCoreWithSDK(
                "a",
                pathContext.SolutionRoot,
                "net472");
            var projectB = SimpleTestProjectContext.CreateNETCoreWithSDK(
                "b",
                pathContext.SolutionRoot,
                "net472");
            var projectC = SimpleTestProjectContext.CreateNETCoreWithSDK(
                "c",
                pathContext.SolutionRoot,
                "net472");

            await SimpleTestPackageUtility.CreateFolderFeedV3Async(
                pathContext.PackageSource,
                PackageSaveMode.Defaultv3,
                new SimpleTestPackageContext("d", "1.0.0"),
                new SimpleTestPackageContext("d", "2.0.0"),
                new SimpleTestPackageContext("d", "3.0.0")
                );

            solution.Projects.Add(projectA);
            solution.Projects.Add(projectB);
            solution.Projects.Add(projectC);
            solution.Create(pathContext.SolutionRoot);

            AddProjectReference(projectA, projectB);
            AddProjectReference(projectB, projectC);
            AddPackageReferenceToProject(projectC);

            CreateDirectoryPackagesPropsWithVersionForPackageD(pathContext, "a", "1.0.0");
            CreateDirectoryPackagesPropsWithVersionForPackageD(pathContext, "b", "3.0.0");
            CreateDirectoryPackagesPropsWithVersionForPackageD(pathContext, "c", "2.0.0");

            var args = new string[] {
                    "restore",
                    solution.SolutionPath,
                    "-Verbosity",
                    "detailed",
                };

            // Act
            var r = CommandRunner.Run(
                Util.GetNuGetExePath(),
                pathContext.WorkingDirectory.Path,
                string.Join(" ", args),
                waitForExit: true);

            // Assert
            r.Success.Should().BeFalse(because: r.AllOutput);
        }

        private static void CreateDirectoryPackagesPropsWithVersionForPackageD(SimpleTestPathContext pathContext, string projectName, string version)
        {
            var directoryPackagesPropsContent =
                           @$"<Project>                    
                        <ItemGroup>
                            <PackageVersion Include=""D"" Version=""{version}"" />
                        </ItemGroup>
                    </Project>";
            File.WriteAllText(Path.Combine(pathContext.SolutionRoot, projectName, $"Directory.Packages.Props"), directoryPackagesPropsContent);
        }

        private static void AddProjectReference(SimpleTestProjectContext from, SimpleTestProjectContext to)
        {
            var xml = from.GetXML();

            ProjectFileUtils.AddProperty(
                xml,
                "ManagePackageVersionsCentrally",
                "true");

            ProjectFileUtils.AddItem(
                            xml,
                            "ProjectReference",
                            to.ProjectPath,
                            NuGetFramework.AnyFramework,
                            new Dictionary<string, string>(),
                            new Dictionary<string, string>());

            xml.Save(from.ProjectPath);
        }
        
        private static void AddPackageReferenceToProject(SimpleTestProjectContext project)
        {
            var xml = project.GetXML();

            ProjectFileUtils.AddProperty(
                xml,
                "ManagePackageVersionsCentrally",
                "true");

            ProjectFileUtils.AddItem(
                            xml,
                            "PackageReference",
                            "d",
                            NuGetFramework.AnyFramework,
                            new Dictionary<string, string>(),
                            new Dictionary<string, string>());

            xml.Save(project.ProjectPath);
        }
```

This test seems to fail with your changes. 

Is there an issue with the changes or is it my test that might be missing something?

marcin-krystianc replied:
@nkolev92 You didn't include implementation for the `AddPackageReferenceToProject`. Can you share it?

nkolev92 replied:
my bad, added now.

marcin-krystianc replied:
Your test didn't set `CentralPackageTransitivePinningEnabled` property (by default it is false).
Here is an updated version of `AddProjectReference` method for which the test doesn't fail anymore:

```
 private static void AddProjectReference(SimpleTestProjectContext from, SimpleTestProjectContext to)
        {
            var xml = from.GetXML();

            ProjectFileUtils.AddProperty(
                xml,
                "ManagePackageVersionsCentrally",
                "true");

            ProjectFileUtils.AddProperty(
                xml,
                "CentralPackageTransitivePinningEnabled",
                "true");

            ProjectFileUtils.AddItem(
                            xml,
                            "ProjectReference",
                            to.ProjectPath,
                            NuGetFramework.AnyFramework,
                            new Dictionary<string, string>(),
                            new Dictionary<string, string>());

            xml.Save(from.ProjectPath);
        }
```

marcin-krystianc replied:
I've added another test based on your approach.

nkolev92 replied:
Perfect. 
Thanks for your patience with this.

nkolev92 reviewed (dismissed):
Thank you for the contribution @marcin-krystianc and apologies for the delays.

marcin-krystianc committed:
Merge remote-tracking branch 'origin/dev' into dev-marcink-20220504-centraltransitive

marcin-krystianc committed:
Add integration test

nkolev92 commented:
Seems like our CI hook is broken. 

cc @zivkan

zivkan commented:
github webhooks are (were?) having an outage: https://www.githubstatus.com/incidents/gvzcw6sd0xxb

nkolev92 commented:
Can you rebase again @marcin-krystianc 

https://github.com/NuGet/NuGet.Client/commit/e9637ed865687c2153da78c6c6385f010277329c didn't make it in your rebase.

nkolev92 commented:
Basically NuGet.CommandLine.Test.NuGetPushCommandTest.PushCommand_InvalidInput_V2HttpSource(invalidInput: "https://invalid-2a0358f1-88f2-48c0-b68a-bb150cac00"...)

 is failing.

nkolev92 reviewed (approved):
LGTM. 

I have a hunch about some other potential bugs we may have with the transitive walking. 

I'll ping everyone on the details if it indeed does turn out to be a bug.

nkolev92 commented:
So the thing I was worried about is that if a package is transitively specified in a child project, the parent project gets a library dependency that's `centrally` managed. 

Turns out it doesn't hurt transitive pinning, but I wonder if it could mess with it in a different way. 

This is the test I used (I basically tweaked the test being added in this change). 

```cs

/// <summary>
        /// A 1.0 -> D 3.0 (Central transitive) ->  E 1.0.0
        ///       -> B 1.0 -> D 2.0.0 (declared)
        ///       -> B 1.0 -> E 2.0.0
        /// </summary>
        [Fact]
        public async Task RestoreNetCore_DependenciesFromRootProjectThatArePinned_ShouldBeIgnored()
        {
            // Arrange
            using var pathContext = new SimpleTestPathContext();

            // Set up solution, project, and packages
            var solution = new SimpleTestSolutionContext(pathContext.SolutionRoot);

            var projectB = CreateProject(pathContext, "B");
            var projectA = CreateProject(pathContext, "A", projectB);

            var packageD200 = new SimpleTestPackageContext("D", "2.0.0");
            var packageD300 = new SimpleTestPackageContext("D", "3.0.0");
            var packageE100 = new SimpleTestPackageContext("E", "1.0.0");
            var packageE200 = new SimpleTestPackageContext("E", "2.0.0");
            packageD300.Dependencies.Add(packageE100);

            await SimpleTestPackageUtility.CreateFolderFeedV3Async(
                pathContext.PackageSource,
                PackageSaveMode.Defaultv3,
                packageD200,
                packageD300,
                packageE100,
                packageE200
            );

            solution.Projects.Add(projectA);
            solution.Projects.Add(projectB);
            solution.Create(pathContext.SolutionRoot);

            AddPackageReferenceToProject(projectB);

            CreateDirectoryPackagesPropsWithVersionForPackageD(pathContext, projectA, "3.0.0");
            CreateDirectoryPackagesPropsWithVersionForPackageDAndE(pathContext, projectB, "2.0.0");

            var args = new string[] {
                    "restore",
                    solution.SolutionPath,
                    "-Verbosity",
                    "detailed",
                };

            // Act
            var r = CommandRunner.Run(
                Util.GetNuGetExePath(),
                pathContext.WorkingDirectory.Path,
                string.Join(" ", args),
                waitForExit: true);

            // Assert
            r.Success.Should().Be(true, because: r.AllOutput);

            var assetsFile = projectA.AssetsFile;

            assetsFile.Libraries.Should().HaveCount(3); // B, D & E.
            var eLibrary = assetsFile.Libraries.Single(e => e.Name.Equals("E"));
            eLibrary.Version.Should().Be(NuGetVersion.Parse("1.0.0"));
            // Local methods
            void CreateDirectoryPackagesPropsWithVersionForPackageD(SimpleTestPathContext pathContext, SimpleTestProjectContext projectContext, string version)
            {
                var directoryPackagesPropsContent =
                    @$"<Project>
                            <ItemGroup>
                                <PackageVersion Include=""D"" Version=""{version}"" />
                            </ItemGroup>
                        </Project>";
                var directoryName = Path.GetDirectoryName(projectContext.ProjectPath);
                File.WriteAllText(Path.Combine(directoryName, $"Directory.Packages.Props"), directoryPackagesPropsContent);
            }

            void CreateDirectoryPackagesPropsWithVersionForPackageDAndE(SimpleTestPathContext pathContext, SimpleTestProjectContext projectContext, string version)
            {
                var directoryPackagesPropsContent =
                    @$"<Project>
                            <ItemGroup>
                                <PackageVersion Include=""D"" Version=""{version}"" />
                                <PackageVersion Include=""E"" Version=""2.0.0"" />
                            </ItemGroup>
                        </Project>";
                var directoryName = Path.GetDirectoryName(projectContext.ProjectPath);
                File.WriteAllText(Path.Combine(directoryName, $"Directory.Packages.Props"), directoryPackagesPropsContent);
            }

            SimpleTestProjectContext CreateProject(SimpleTestPathContext pathContext, string name, SimpleTestProjectContext referencedProject = null)
            {
                var projectContext = SimpleTestProjectContext.CreateNETCoreWithSDK(
                    name,
                    pathContext.SolutionRoot,
                    "net472");

                projectContext.Properties.Add("ManagePackageVersionsCentrally", "true");
                projectContext.Properties.Add("CentralPackageTransitivePinningEnabled", "true");

                if (referencedProject != null)
                    projectContext.AddProjectToAllFrameworks(referencedProject);

                return projectContext;
            }

            void AddPackageReferenceToProject(SimpleTestProjectContext project)
            {
                var xml = project.GetXML();

                ProjectFileUtils.AddItem(
                    xml,
                    "PackageReference",
                    "D",
                    NuGetFramework.AnyFramework,
                    new Dictionary<string, string>(),
                    new Dictionary<string, string>());

                xml.Save(project.ProjectPath);
            }
        }
```

marcin-krystianc committed:
Merge remote-tracking branch 'origin/dev' into dev-marcink-20220504-centraltransitive

marcin-krystianc commented:
@nkolev92 I've merged dev branch into mine. 

> I have a hunch about some other potential bugs we may have with the transitive walking.

Yeah, it is quite messy if you start considering graphs in which different projects have a different sets of central package versions and both use transitive pinning. That is definitely not a common scenario anyway.
I hope that my change removes all the ambiguity. The transitive pining information is going to be respected only for the entry project and ignored for other projects, so it is much simpler to reason about it. 
This change also enables some performance improvements in the graph walking algorithm when transitive pinning is enabled  (we can do some dictionary lookups instead of walking the entire graph).

ghost commented:
This PR has been automatically marked as stale because it has no activity for **30 days**. It will be closed if no further activity occurs **within another 60 days** of this comment. If it is closed, you may reopen it anytime when you're ready again, as long as you don't delete the branch.

nkolev92 commented:
@jeffkl's out and he'll handle it when back. 
We expect this change to go into 7.0.100 of the SDK, 6.4 of NuGet, 17.4 of VS.

It was pretty late in the game for 6.3 so we didn't feel comfortable merging.

marcin-krystianc committed:
Merge remote-tracking branch 'origin/dev' into dev-marcink-20220504-centraltransitive

jeffkl reviewed (approved):
Thanks for the contribution @marcin-krystianc !
